### PR TITLE
Add cached-copy localization strategy.

### PR DESCRIPTION
**Pull request by rhpvorderman**
### The problem

Containers are great, but soft-links cannot be used. Hard-links in the execution folder are fine, but these cannot be used across physical disks. Most HPC have a filesystem consisting of multiple disks. This means only `copy` can be used as a localization strategy.

Say you want to align multiple samples to the reference genome. And the reference genome is on another physical disk. That means the reference genome is copied once for each sample. That takes a lot of IO time and a lot of physical disk space. This makes cromwell practically unusable for use with containers on a HPC.  Multiple people have the same problem and proposed a solution in #2620, #3447 and #4711. 

### This solution

This PR aims to tackle the problem by adding a new strategy called `cached-copy`. Since hard-links do not work across physical disks and soft-links do not work, copying is unavoidable. However, instead of copying the file directly into `<call_dir>/inputs` the file is first copied to a cache in `<cromwell_root>/cached-inputs`. This should ensure that `cached-inputs` is on the same physical disk as `<call_dir>` (which is also in `<cromwell_root>`). The cached file is then hard-linked to `<call_dir>/inputs`. When the same file is needed again it will check if the file already exists in the `cached-inputs` and then hard-link it. This saves all the time needed for copying.

### Technical notes
File collisions are prevented by appending the last modified time epoch millisecond to the name of the cached file. If the input file is changed between jobs it will create a new cache entry. Md5sums are not used because it is extremely compute intensive for large files. Path+modtime should guarantee that  files are the same. 

I have expanded the SFS test scala file so it properly tests the new `cached-inputs` strategy.

I have added information on how to use the strategy in the docs, and added this PR to the changelog.

### Help still needed

There are two things that I could not figure out without cromwell developer help:

~~* Checking whether a file exists and copying it to the cache should never be done by multiple threads simeltaneously. I have used the `synchronized` method to prevent this. I used an object for this, because I am sure it is unique within the JVM at cromwell runtime. This works fine, but I can imagine this can be solved in a nicer way using akka? However the akka documentation is an extensive jungle on its own, and requires quite some expertise to navigate. I could not find very quickly what I needed, and the `synchronization` primitive works fine. It is also **just 2 lines of extra code**. So if the akka solution is quite elegant as well I would like to learn about that. If not, well, it is not too bad having 2 lines of understandable commented code that is not "the proper way of doing things(TM)".~~

* I used the SFS scalatests to make sure everything worked correctly. However this did not test whether the thread safety was working correctly. I have added a test wdl in centaur: `standardTestCases/cached_copy/cached_copy.wdl`. This workflow creates 10 jobs that read the same input file. This workflow will crash if the `cached-inputs` cache is not used in a thread-safe way. I tested this manually with `java -Dbackend.providers.Local.config.filesystems.local.localization.0="cached-copy" -jar server/target/scala-2.12/cromwell-41-*-SNAP.jar  run centaur/src/main/resources/standardTestCases/cached_copy/cached_copy.wdl` . Is there a way to integrate such a test in scalatest file? I have tried the `.par` method, but that did not quite work.

I hope you will consider this PR as it solves an important issue for us. Thanks!

### CODE REVIEW

**Comment by mcovarr** [comment]
Thank you for the PR! The Cromwell team is unfortunately a bit slammed this week but we hope to get back to you with more substantial feedback next week. 🙂

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala [hunk 1]
```diff
@@ -99,6 +99,37 @@ trait SharedFileSystem extends PathFactory {
...
+      // Md5 is safer but much much slower and way too CPU intensive for big files.
+      val pathAndModTime: String = originalPath.lastModifiedTime.toEpochMilli.toString + originalPath.name
+      val cachedCopyPath: Path = cachedCopySubDir./(pathAndModTime)
+
+      // The following step cannot be performed multithreaded. Checking if the file exists and copying of the file should never
+      // happen simultaneously. Therefore this is synchronized using an object that only exists once in the JVM.
+      // This is probably not the proper 'akka' way of doing things. But I could not find the proper way of doing things
+      // after an extensive web search and read through the akka documentation.
+
+      SharedFileSystem.synchronized {
```
Comment (aednichols) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: While looking over this PR, we're a bit scared of this global lock... have you measured the impact to performance?

It would be nice to see a before/after measurement for workflow execution.

Comment (aednichols) [reply supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: Also, we are plotting a course towards multiple Cromwells running jobs simultaneously from the same DB ("Horicromtal"), and this lock mechanism will not work across JVMs.

Comment (rhpvorderman) [reply supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: > While looking over this PR, we're a bit scared of this global lock... have you measured the impact to performance?

I am not too happy with this either. Performance-wise it is much better than repeated copying! But on the other hand this limits file localization to one thread. 
I would love to have this thread lock by based on the filepath, so then it only locks when it gets requests for the same file (which is good) but not one another file is copied.

I haven't figured out to do this, and I was wondering if you have any suggestions?

Comment (rhpvorderman) [reply supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: > Also, we are plotting a course towards multiple Cromwells running jobs simultaneously from the same DB ("Horicromtal"), and this lock mechanism will not work across JVMs.
Thanks for letting me know.
Hmm `.lock` files are to slow. So this is going to be difficult. Maybe we can use the database?

Comment (rhpvorderman) [reply supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: @aednichols I have put some thought to this and have come to the following conclusions:

1. The global lock is okay. As long as it does not include the actual copying step. The performance impact of a few instructions/clock cycles spent in a lock is minimal.
Performance-wise cached-copy is much better than repeated copying. But it is a shame that localization of `<BWA index>` has to wait for the copying of `<Reference genome>`. These files shouldn't wait on eachother.
2. `Horicromtal` is only going to be a problem when these instances share the same execution folder. But this is not a far-fetched use-case. So:
  * Communicating between cromwell processes via the database will make the code much more complicated and harder to maintain. It will also complicate the database structure.
  * The filesystem is not fast enough for locking files **within** a cromwell process, because many localizations in scatters are mostly started at almost the same time. However between processes the chance that two cromwell processes need the same file within the interval it takes to create a lock file is negligible.

So in conclusion: I am going to change the code so the locks will be based on a per originalPath bases. Between cromwellprocesses lock files are fast enough, so these will be implemented. Within cromwell processes this is not fast enough. An in-memory dictionary will be used to keep track of paths that are being copied.

**Comment by rhpvorderman** [comment]
@aednichols . I implemented the changes. The lock should have a negligible performance impact. It only locks when cached-copy strategy is used, threads which use other strategies are not blocked by this. (This was already true before the changes).

A dictionary is used to keep track which files are being copied within the process. Filesystem lock files are too slow for this because during scatters cromwell creates a lot of threads that need exactly the same file at the same time.  

A lock file is now used so other cromwell processes know the file is being copied. The chances of two processes needing the same file at exactly the same time is negligible, so lock files are fast enough here.

Due to the `synchronized` lock, there can not be race conditions where more than one thread modifies the dictionary and/or creates the lock file at the same time. (The lock is absolutely necessary for cached-copy to work). The amount of time spent in the lock per thread is negligible. 

I implemented a waitOnCopy function which lets a thread wait for the locks in the dictionary on the filesystem to clear. This structure allows a lot of paths to be copied to the cache at the same time. So the copying is still multithreaded, maintaining high performance.

Are these changes satisfactory?

**Comment by aednichols** [comment]
Thanks for the update! We will assign reviewers to give proper feedback, as soon as someone is available.

**Comment by rhpvorderman** [comment]
@gemmalam Could you remove the "back with originator" label? This PR has already returned from the originator. I am waiting for feedback. So if this stays with "back with originator" we will be in waiting on eachother deadlock :slightly_smiling_face: .

**Comment by rhpvorderman** [comment]
> Thanks for the update! We will assign reviewers to give proper feedback, as soon as someone is available.

@aednichols could you give an indication when the feedback will be coming?

**Comment by aednichols** [comment]
There's quite a bit of debate internally about this PR. Some team members remain deeply uncomfortable with how locking is handled, but it would take us a lot of time to research and recommend a better solution.

**Comment by rhpvorderman** [comment]
I can understand. This solution is not ideal. On the upside: it is only activated for those who willingly put "cached-copy" in their configs. The rest of the cromwell users are **not** affected by the lock mechanism. By default this does **not** affect anyone. I could adapt this PR and plaster the words: `WARNING: EXPERIMENTAL` all over it if that helps.

**Comment by rhpvorderman** [comment]
Also. File locks are not that good. But locking via the database would be ideal for horicromtal. If I get some pointers I can implement a database lock. This will require an extra table or something, so I need some pointers on how this should be ideally integrated in cromwell.

**Comment by rhpvorderman** [comment]
> There's quite a bit of debate internally about this PR. Some team members remain deeply uncomfortable with how locking is handled, but it would take us a lot of time to research and recommend a better solution.

 If I may reiterate: by default this does not break anything for anyone. The locking only happens when `cached-copy` is set in the config consciously by the user. I maintain [my own patched jar for cromwell](https://github.com/rhpvorderman/cromwell/releases/tag/41-LUMC-patches), because this is taking very long already. We are running this in production and not experiencing problems.

(There are only problems when the maximum number of hardlinks is reached, then cromwell defaults to copying again. It does not break anything, but it will use a lot of space on the filesystem and it will slowdown pipeline runs. I am working on a fix for that as well.)

File: docs/backends/HPC.md [hunk 2]
```diff
@@ -18,11 +18,20 @@ Cromwell is configured with a root execution directory which is set in the confi
 When Cromwell runs a workflow, it first creates a directory `<cromwell_root>/<workflow_uuid>`.  This is called the `workflow_root` and it is the root directory for all activity in this workflow.
 
 Each `call` has its own subdirectory located at `<workflow_root>/call-<call_name>`.  This is the `<call_dir>`.
-Any input files to a call need to be localized into the `<call_dir>/inputs` directory. There are different localization strategies that Cromwell will try until one works. Below is the default order specified in `reference.conf` but it can be overridden:
+Any input files to a call need to be localized into the `<call_dir>/inputs` directory. There are different localization strategies that Cromwell will try until one works:
 
 * `hard-link` - This will create a hard link to the file
 * `soft-link` - Create a symbolic link to the file. This strategy is not applicable for tasks which specify a Docker image and will be ignored.
 * `copy` - Make a copy the file
+* `cached-copy` introduced in cromwell 41 as an experimental feature. This copies files to a file cache in
```
Comment (aednichols) [inline docs/backends/HPC.md]: Version number needs updating (or removing, IMO people can find the version from the changelog)

File: docs/backends/HPC.md [hunk 3]
```diff
@@ -18,11 +18,20 @@ Cromwell is configured with a root execution directory which is set in the confi
...
 * `copy` - Make a copy the file
+* `cached-copy` introduced in cromwell 41 as an experimental feature. This copies files to a file cache in 
+`<workflow_root>/cached-inputs` and then hard links them in the `<call_dir>/inputs` directory. 
+
+`cached-copy` is intended for a shared filesystem that runs on multiple physical disks, where docker containers are used. 
+Hard-links don't work between different physical disks and soft-links don't work with docker. Copying uses a lot of
+space if a multitude of tasks use the same input. `cached-copy` copies the file only once to the physical disk containing
+the `<workflow_root>` and then uses hard links for every task that needs the input file. This can save a lot of space.
+
+The default order in `reference.conf` is `hard-link`, `soft-link`, `copy`
```
Comment (aednichols) [inline docs/backends/HPC.md]: Is `cached-copy` not in the list?

File: docs/backends/HPC.md (hunk 2 above)
Comment (mcovarr) [reply docs/backends/HPC.md]: Yes please remove the version number.

File: docs/backends/HPC.md (hunk 3 above)
Comment (mcovarr) [reply docs/backends/HPC.md]: This sounds right, cached-copy is an experimental feature and should not be on the default list.

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala [hunk 4]
```diff
@@ -122,6 +207,7 @@ trait SharedFileSystem extends PathFactory {
     // If localizing for a docker job, remove soft-link as an option
```
Comment (mcovarr) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystem.scala]: Please update the comment to describe the action being taken here for `cached-copy` as well.

File: supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystemJobCachingActorHelper.scala [hunk 5]
```diff
@@ -19,5 +19,11 @@ trait SharedFileSystemJobCachingActorHelper extends StandardCachingActorHelper {
     override lazy val sharedFileSystemConfig: Config = {
       configurationDescriptor.backendConfig.as[Option[Config]]("filesystems.local").getOrElse(ConfigFactory.empty())
     }
+
+    // cachedCopyDir should be on the same physical filesystem as the execution root.
+    // WDL workflow names may not contain '-' so using 'cached-inputs' will certainly
+    // not collide with any workflows in the root directory.
+    override lazy val cachedCopyDir: Option[Path] = Some(
```
Comment (mcovarr) [inline supportedBackends/sfs/src/main/scala/cromwell/backend/sfs/SharedFileSystemJobCachingActorHelper.scala]: Please use `Option` rather than `Some` throughout these changes to construct `Option` instances. This is the convention in Cromwell for null-safety even if it doesn't actually make a difference in this particular case. 😉

File: docs/backends/HPC.md (hunk 3 above)
Comment (rhpvorderman) [reply docs/backends/HPC.md]: This is indeed the reason I did not put it there. The strategy has some hairiness to it that the other strategies don't have. The others never have to take race conditions into account. Race conditions can cause weird errors and stability issues. While I believe I fixed the race conditions, this can never be 100% guaranteed.

**Comment by rhpvorderman** [comment]
@aednichols  and @mcovarr Thank you very much for the reviews. @aednichols Thank you very much for keeping me informed about the status of this PR, it is much appreciated.

The code has been updated to reflect the comments. I also merged the latest branch of develop in and updated the changelog.

**Comment by rhpvorderman** [comment]
Two errors on the testing:
```
- should successfully run curl *** FAILED *** (1 minute, 37 seconds)
  centaur.test.CentaurTestException: Unexpected terminal status Failed but was waiting for Succeeded (workflow ID: bb88b541-3f1a-490c-9121-7685b4ab54b3). Metadata 'failures' content: [
  {
    "causedBy" : [
      {
        "causedBy" : [
        ],
        "message" : "Job curl_wf.newsgrab:NA:1 exited with return code 6 which has not been declared as a valid return code. See 'continueOnReturnCode' runtime attribute for more details."
      }
    ],
    "message" : "Workflow failed"
  }
]
```

```
info] PubSubMetadataServiceActorSpec:
[info] A PubSubMetadataActor with a subscription should 
[info] - should create the requested subscription *** FAILED *** (17 milliseconds)
[info]   java.lang.AssertionError: received 1 excess messages on InfoFilter(None,Left(Creating subscription baz),true)
[info]   at akka.testkit.EventFilter.intercept(TestEventListener.scala:116)
[info]   at cromwell.services.metadata.impl.pubsub.PubSubMetadataServiceActorSpec.$anonfun$new$9(PubSubMetadataServiceActorSpec.scala:40)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[info]   at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
[info]   at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
```
I don't see how these are caused by this PR. I would gladly fix them if I would know how.

**Comment by mcovarr** [comment]
Those two failures are probably unrelated test flakiness (the curl exit 6 is a "Couldn't resolve host" error). I've restarted the two builds, hopefully the errors will go away.

**Comment by illusional** [comment]
Hey @rhpvorderman, I've started to use this for our workflows and seems to be working well! Props for this change :)

 I've got a small suggestion (not enough to raise an issue, and only if you're already making other changes), it would be awesome if Cromwell could log a message to say that it's copying files. I watch for that because then I know the task is starting properly. 

Unrelated to that, I was wondering what hurdles might have to be overcome to devise a hashing-strategy based on your new `cached-copy` (that's not File / md5).

**Comment by rhpvorderman** [comment]
@illusional 
I am happy you like this change. I have checked your other post in #4945 and your use case is similar to ours. We use a SGE cluster and run cromwell from the headnode.

The message is really easy to implement. But I am not sure what would be the right way to tackle this. I would like some consistency with the other localization methods, and I don't know if they message when a file is being copied.

I haven't tested cached-copy in conjunction with call-caching and path+modtime yet. If I find issues with it I will create a new issue on the cromwell issue tracker, ping you, and see if I can fix it in a PR. We rely heavily on the path+modtime strategy as well.

**Comment by illusional** [comment]
Ooop, I might have misspoke. I thought the copy strategy did actually log that it was copying, but I realised that what I was seeing was that the `hard link` had failed, and knew that it was copying based on that:

> `WARN  - Localization via hard link has failed: /path/to/destination.file -> /path/to/original.file: Invalid cross-device link`

I think it still might be useful, but I realise there's actually no precedent here.

---

Oh, so the path+modtime sort of just works? I was under the impression it wouldn't for those cache-strategies. I don't know if it wouldn't try, or would never succeed because I never tried, but here's what the [docs say](https://cromwell.readthedocs.io/en/stable/Configuring/#local-filesystem-options):

> - "path" will compute an md5 hash of the file path. This strategy will only be effective if the duplication-strategy (above) is set to "soft-link",
> - "path+modtime" will compute an md5 hash of the file path and the last modified time. The same conditions as for "path" apply here.

Thanks for the reply!

**Comment by illusional** [comment]
Hey @rhpvorderman, just wanted to check in and say this has been working great!

I was thinking about adding a new one, called "cached-hardlink" or so, but instead of always copying, it would try to hard-link it first. We have the use case where often our bigger sequence data is on the same disk, but some reference files are not. My plan was to just follow your pattern, but use a modified `localizePathViaCachedCopy` (like `localizePathViaHardlink`). Do you have any advice if I were to give it a crack over the EOY break?

**Comment by rhpvorderman** [comment]
@illusional You can set the priorities for localization like this
```
            localization: [
              "hard-link", "cached-copy", "copy"
            ]
```
In that case it will try to hard-link first. It works on our setup.

> Hey @rhpvorderman, just wanted to check in and say this has been working great!

Thanks for letting us know. In our institute it is also working great. We have had zero crashes, deadlocks etc. related to this code. So we are also quite happy with it. I am very glad it is useful to others as well.

**Comment by illusional** [comment]
I think this might've been broken in Cromwell-51, I've raised a new ticket: #5533

**Comment by rhpvorderman** [comment]
Thanks for the heads up. We haven't gotten around to move our cluster to 51 yet. Pinging @DavyCats so he is also aware.
We will investigate the issue.

**Comment by rhpvorderman** [comment]
@illusional @DavyCats I have confirmed the problem. This is pretty serious. I will see what I can do.

**Comment by rhpvorderman** [comment]
Scratch that. I just saw the solution in https://github.com/broadinstitute/cromwell/issues/5533
Thanks for raising this issue!

**Comment by illusional** [comment]
Haha yeah I was just about to say I think it's working with a revised config change. I PRd a doc change, but it's only in the [develop docs](https://cromwell.readthedocs.io/en/develop/Configuring/#local-filesystem-options).

**Comment by rhpvorderman** [comment]
Good work! For me the problem was caused because I use the conda installation of cromwell and I made a typo in the `_JAVA_OPTIONS` environment variable, so it was running locally instead of on the cluster :man_facepalming: . Still I managed to find the issue because of your issue :smile: .
